pkg/solana/txm/pendingtx.go

```diff
@@ -563,6 +576,52 @@ func (c *pendingTxContext) TrimFinalizedErroredTxs() int {
 	return len(expiredIDs)
 }
 
+func (c *pendingTxContext) IsTxReorged(sig solana.Signature, sigOnChainState txmutils.TxState) (string, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	// Grab in memory state of the signature
+	txInfo, exists := c.sigToTxInfo[sig]
+	if !exists {
+		return "", false
+	}
+
+	// Compare our in-memory state of the sig with the current on-chain state to determine if the sig had a regression
+	sigInMemoryState := txInfo.state
+	var hasReorg bool
+	switch sigInMemoryState {
+	case txmutils.Confirmed:
+		if sigOnChainState == txmutils.Processed || sigOnChainState == txmutils.Broadcasted || sigOnChainState == txmutils.NotFound {
+			hasReorg = true
+		}
+	case txmutils.Processed:
+		if sigOnChainState == txmutils.Broadcasted || sigOnChainState == txmutils.NotFound {
+			hasReorg = true
+		}
+	default: // No reorg if the signature is not in a state that can be reorged
+	}
+
+	return txInfo.id, hasReorg
+}
+
+func (c *pendingTxContext) GetPendingTx(id string) (pendingTx, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	var tx, tempTx pendingTx
+	var broadcastedExists, confirmedExists bool
+	if tempTx, broadcastedExists = c.broadcastedProcessedTxs[id]; broadcastedExists {
+		tx = tempTx
+	}
+	if tempTx, confirmedExists = c.confirmedTxs[id]; confirmedExists {
+		tx = tempTx
+	}
+
```

Comment:
Shouldn't it be impossible for a tx to exist in both confirmed and broadcasted since it's deleted during `OnCofirmed`? Not saying this is a huge issue mostly just testing my understanding

Reply:
Yes, this shouldn't happen. What is your concern about this?

Reply:
Well this function seems to be okay with that case where `broadcastedExists` and `confirmedExists`, I guess the reorg would just be caught elsewhere?

Reply:
We catch the reorgs here: 

- When [getting a NotFound](https://github.com/smartcontractkit/chainlink-solana/blob/ea4d0fa7633b7a714c38ddf38b1872f61c6dbfe5/pkg/solana/txm/txm.go#L443) sig, we check if our in-memory state is higher for it (`Processed` or `Confirmed`) and needs to be handled
- When [getting a Processed signature](https://github.com/smartcontractkit/chainlink-solana/blob/ea4d0fa7633b7a714c38ddf38b1872f61c6dbfe5/pkg/solana/txm/txm.go#L458) sig, if our in-memory state is higher for it (`Confirmed`) and needs to be handled
- By definition, no reorgs from `Finalized` state should happen, that's why they are not handled